project/project_repository_blackbox_test.go

```diff
@@ -0,0 +1,173 @@
+package project_test
+
+import (
+	"testing"
+
+	"github.com/almighty/almighty-core/errors"
+	"github.com/almighty/almighty-core/gormsupport"
+	"github.com/almighty/almighty-core/project"
+	satoriuuid "github.com/satori/go.uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/net/context"
+)
+
+var testProject string = satoriuuid.NewV4().String()
+var testProject2 string = satoriuuid.NewV4().String()
+
+func TestRunProjectRepoBBTest(t *testing.T) {
+	suite.Run(t, &projectRepoBBTest{DBTestSuite: gormsupport.NewDBTestSuite("../config.yaml")})
+}
+
+type projectRepoBBTest struct {
+	gormsupport.DBTestSuite
+	undoScript *gormsupport.DBScript
+	repo       *project.UndoableProjectRepository
+}
+
+func (test *projectRepoBBTest) SetupTest() {
+	test.undoScript = &gormsupport.DBScript{}
+	test.repo = project.NewUndoableProjectRepository(project.NewProjectRepository(test.DB), test.undoScript)
+	test.DB.Unscoped().Delete(&project.Project{}, "Name=?", testProject)
+	test.DB.Unscoped().Delete(&project.Project{}, "Name=?", testProject2)
+}
+
+func (test *projectRepoBBTest) TearDownTest() {
+	test.undoScript.Run(test.DB)
+}
+
+func (test *projectRepoBBTest) TestCreate() {
+	res, _ := expectProject(test.create(testProject), test.requireOk)
+
+	require.Equal(test.T(), res.Name, testProject)
+
+	expectProject(test.create(""), test.assertBadParameter())
+	expectProject(test.create(testProject), test.assertBadParameter())
+}
+
+func (test *projectRepoBBTest) TestLoad() {
+	expectProject(test.load(satoriuuid.NewV4()), test.assertNotFound())
+	res, _ := expectProject(test.create(testProject), test.requireOk)
+
+	res2, _ := expectProject(test.load(res.ID), test.requireOk)
+	assert.True(test.T(), (*res).Equal(*res2))
+}
+
+func (test *projectRepoBBTest) TestSaveOk() {
+	res, _ := expectProject(test.create(testProject), test.requireOk)
+
+	newName := satoriuuid.NewV4().String()
+	res.Name = newName
+	res2, _ := expectProject(test.save(*res), test.requireOk)
+	assert.Equal(test.T(), newName, res2.Name)
+}
+
+func (test *projectRepoBBTest) TestSaveFail() {
+	p1, _ := expectProject(test.create(testProject), test.requireOk)
```

Comment:
Having a complete test on the lower db layer makes it easier to eliminate possible causes of failure in the REST controller since the higher level probably have more logic than just simple DB Crud.